File: 02-course/07-树/144.go
package main

/**
 * Definition for a binary tree node.
 */

// 参考：https://leetcode-cn.com/problems/binary-tree-preorder-traversal/solution/jian-dan-yi-dong-javac-pythonjs-er-cha-s-3vyq/

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func preorderTraversal(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}
	var stack []*TreeNode
	var res []int

	stack = append(stack, root)

	for len(stack) > 0 {
		cur := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		res = append(res, cur.Val)

		if cur.Right != nil {
			stack = append(stack, cur.Right)
		}
		if cur.Left != nil {
			stack = append(stack, cur.Left)
		}
	}
	return res
}

File: 刷题/二叉树/144.py
# Definition for a binary tree node.
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# 前序遍历
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []
        stack, res = [], []

        while len(stack) > 0 or root:
            while root:
                res.append(root.val)
                stack.append(root)
                root = root.left

            root = stack.pop()
            root = root.right
        return res
